dvc/repo/experiments/executor/base.py

```diff
@@ -464,6 +466,20 @@ def checkpoint_callback(
             exp_rev = cls.commit(
                 scm, exp_hash, exp_name=name, force=force, checkpoint=True
             )
+
+            git_remote = os.environ.get("DVC_EXP_AUTO_PUSH", None)
```

Comment:
Please put this env var into `dvc/env.py` and use it as a var.

Reply:
what about a different meaning, e.g. `DVC_EXP_AUTO_PUSH=10` meaning "push every 10 checkpoints?"

Reply:
@casperdcl 
`DVC_EXP_AUTO_PUSH` is the remote git repo for auto push. Maybe this name needs to be modified. But it is better to discuss this problem in the PR for `dvc.org`

Reply:
> PR for `dvc.org`

I don't follow. I understand what you're using `DVC_EXP_AUTO_PUSH` for. I'm asking for:

- `DVC_EXP_AUTO_PUSH` to be changed to mean something else (push every N epochs)
- or a new variable `DVC_EXP_AUTO_PUSH_EVERY` added (push every N epochs)
- or 2 different variables with less ambiguous names `DVC_EXP_PUSH_CHECKPOINT=0`, `DVC_EXP_PUSH_REMOTE=""`

Basically I think a "push every N epochs" config should be supported in this PR rather than in a follow-up so that we can keep API consistent. WDYT @efiop?

Reply:
Is "push every N commits" actually a requirement for CML? This seems like an additional feature request that's unrelated to being able to recover checkpoints from a dead CI runner.

And it seems like there's still some behavior discussion that would need to be worked out there, i.e. what happens if I have it set to 10 and my experiment generates less than 10 commits, should DVC not push anything?

It also seems like something that should maybe be tied to dvclive and not DVC itself, since DVC doesn't even have a real concept of epoch count/iteration awareness (related dvclive ticket: https://github.com/iterative/dvclive/issues/113). Since dvclive *is* aware of actual epoch counts, it can just call `experiments.push` at the appropriate time without the need for adding more env vars

Reply:
@pmrowla 
> And it seems like there's still some behavior discussion that would need to be worked out there, i.e. what happens if I have it set to 10 and my experiment generates less than 10 commits, should DVC not push anything?

The final push after commit is a solution, I think.

Reply:
With regard to this PR, the `DVC_EXP_AUTO_PUSH` env var is intended to be used when there is a requirement that all commits be pushed as they are made (so for CI and other remote execution environments where the user is concerned about potential data loss)

I think granular control over DVC behavior relative to metric states/epoch counts/etc belongs in dvclive. Similar to the discussion over telling dvclive to stop iterating once some metric condition or epoch count is reached, it would be useful for users to be able to push on those kind of conditions as well via dvclive (something like `dvclive.next_step(push=...)` was even suggested in the original issue for this PR).

But this all seems unrelated to the immediate CML need for automated pushing on every commit to account for potentially dead runners (and the DVC fix for this CML requirement should not be dependent on dvclive), which is addressed by what's in this PR now

Reply:
> It also seems like something that should maybe be tied to dvclive and not DVC itself, since DVC doesn't even have a real concept of epoch count/iteration awareness (related dvclive ticket: [iterative/dvclive#113](https://github.com/iterative/dvclive/issues/113)). Since dvclive _is_ aware of actual epoch counts, it can just call `experiments.push` at the appropriate time without the need for adding more env vars

Agreed

Reply:
> I think granular control over DVC behavior relative to metric states/epoch counts/etc belongs in dvclive. Similar to the discussion over telling dvclive to stop iterating once some metric condition or epoch count is reached, it would be useful for users to be able to push on those kind of conditions as well via dvclive (something like `dvclive.next_step(push=...)` was even suggested in the original issue for this PR).

How about let the dvclive control `DVC_EXP_AUTO_PUSH` env dynamically?

Reply:
I don't really see why dvclive would need to manipulate the env var when it can just call the push function directly itself and keep control over the specific push options.

(In CML to fix the dead runner issue we need to push all 3 of commits/dvc cache/run cache, but in normal user scenarios they probably do not need to push all 3 every time)

Reply:
> Is "push every N commits" actually a requirement for CML?

A stable API is a requirement. My point is if we (in this PR or in a future PR) implement "push every N commits" and other config options we'd probably change the API of this PR.

In summary, as a minimum requirement to approve this PR, I would rename `DVC_EXP_AUTO_PUSH` to `DVC_EXP_AUTOPUSH_REMOTE` or something similar.

Reply:
So your only concern for now is the name of this env var @casperdcl?

Seems like `DVC_EXP_AUTO_PUSH` does two things:
1. Flags whether to auto push.
2. Specifies the remote.

The name makes sense for the first purpose but is probably too generic for the second. We could:
1. Keep it as is.
2. Change the name to mention `REMOTE` or even `GIT_REMOTE`/`GITREMOTE`/`GIT`.
3. Split into two env vars.
4. Leave the remote to be configured in Git (see https://git-scm.com/docs/git-config#Documentation/git-config.txt-remotepushDefault) and make `DVC_EXP_AUTO_PUSH` a boolean.

Any preferences?

Reply:
I don't think relying on the default git remote is a solution since it is probably not set in CI environments. I'd be in favor of just changing it to something with `GIT_REMOTE` or `GIT_URL`.

Reply:
`GIT_REMOTE` didn't indicate anything about `DVC` and `AUTOPUSH`, maybe `DVC_CHECKPOINT_REMOTE` is better?

Reply:
We meant that the name should include something like `GIT_REMOTE`/`GIT_URL`, not that the env var should just be `GIT_REMOTE`. So something like `DVC_EXP_AUTO_PUSH_GIT_REMOTE`, but preferably shorter.

I don't think using just `REMOTE` alone will work here since it's not obvious if it's a git remote or a dvc remote.

This is also not specific to checkpoints, regular experiments will end be pushed the same way if the env var is set.

Reply:
> So your only concern for now is the name of this env var?

only blocking concern, yes (because *APIs are forever*).

I'd suggest:

- `DVC_EXP_CHECKPOINT_PUSH:bool`
- rely on Git remote (if CI envs don't set it, we can document that the user should `git remote add` etc. first)

introducing `DVC_EXP_GIT_REMOTE` to me seems necessary [iff](https://en.wikipedia.org/wiki/If_and_only_if) there's a reason for the user to deliberately configure multiple different Git remotes.

Reply:
Configuring multiple git remotes is a pretty typical use case (original/`upstream` repo vs my fork/`origin`). If I'm working on a fork and want my run experiments pushed to the original repo but not my fork, I need to be able to configure the remote.

---

I'm also concerned that relying on the default git remote behavior is going to end up confusing for users in certain situations.

Default git remotes only apply to git branches. When I have a checked out git branch, and do
```
git push
```
with no remote specified, git will look at the repo config and checks for `remote.pushDefault` (never set by default in any git repo), and then `branch.<current_branch_name>.remote` (set when you configure a specific remote-tracking branch with `git push -u ...`) in that order. If neither of those are set, it will fallback to `origin`.

The issue here is that upstream tracking branches only apply to branches. DVC exps do not have upstream tracking branches. This means that "default" for DVC experiments is always going to be either `remote.pushDefault` or `origin`.

This is going to be confusing for users in the event that they are running experiments based on a branch with a configured non-`origin` upstream tracking remote. The end result is that users get this behavior:

```
git checkout mybranch  # checks out mybranch
git push -u upstream mybranch  # sets default push remote for mybranch to 'upstream'
DVC_EXP_AUTO_PUSH=1 dvc exp run  # runs an experiment and auto-pushes it to 'origin'
```

In this case it seems like what the user probably wanted (and probably would expect) is that the experiment would be pushed to `upstream`. But the exp ref is actually pushed to `origin`, since `branch.mybranch.remote = upstream` specifically only applies when git is pushing `refs/heads/mybranch` and does not apply to anything in any custom ref namespace.

This ambiguity is why `dvc exp push` always requires an explicit remote name.

Reply:
I think that the behavior of `auto push` should go with `dvc exp push`.  Make different protocols for them will confuse users. If `dvc exp push` can run with a default remote, `auto push` should also support it, vice versa.

Reply:
One possible solution is that 
1. `DVC_EXP_GIT_REMOTE` is used for the default exp push git remote. Both `dvc exp push` and `auto push` can use it (But the auto push cann't work without it, while `dvc exp push` can run git remote arguments).
2. `DVC_EXP_CHECKPOINT_PUSH`/ `DVC_EXP_CHECKPOINT_AUTO_PUSH` can be used for the open/close of auto_push

For the problem of `push every N epochs ` better to discuss with `dvclive` side?

Reply:
> * `DVC_EXP_GIT_REMOTE` is used for the default exp push git remote. Both `dvc exp push` and `auto push` can use it (But the auto push cann't work without it, while `dvc exp push` can run git remote arguments).

But it is a totally new feature, involving changes on `dvc exp push` should not be mixed in this PR.

Reply:
> Configuring multiple git remotes is a pretty typical use case

ok in that case I'd go with

> 2. Split into two env vars

i.e. `DVC_EXP_CHECKPOINT_PUSH:bool` and `DVC_EXP_GIT_REMOTE:str`?

Regarding:

> I think that the behavior of `auto push` should go with `dvc exp push`

This also makes sense to me

Reply:
> This is also not specific to checkpoints, regular experiments will end be pushed the same way if the env var is set.

I think we need to keep `CHECKPOINT` out of the names of these env vars for this reason.

---

Splitting into two env vars makes sense, but it does make the UX more complex. I think it's worth making sure this is really necessary before creating extra complexity for hypothetical future needs, so I have a couple follow-ups:

Why not rename the env var to `DVC_EXP_GIT_REMOTE` and make no other changes? That provides future flexibility but also keeps the UX concise.

Alternatively, I think we could reconsider having a default remote.

> Configuring multiple git remotes is a pretty typical use case (original/`upstream` repo vs my fork/`origin`). If I'm working on a fork and want my run experiments pushed to the original repo but not my fork, I need to be able to configure the remote.

1. Is this likely in CI?
2. Even locally, are users likely to do `git push -u upstream mybranch`? My typical workflow would be to pull from upstream and push to origin by default.
3. If we aren't following Git default remote configuration, why not document that `origin` is the default instead of requiring it to be explicit?

Reply:
> Is this likely in CI?

in CI probably not

> Even locally, are users likely to do git push -u upstream mybranch? My typical workflow would be to pull from upstream and push to origin by default.

I do this regularly with DVC - we have CI configured to run a different set of actions on pull requests vs branches pushed directly to `iterative/dvc`, so there are times when I want to push a certain branch for specific features (that will require the stricter branch tests) to upstream rather than my personal fork

> If we aren't following Git default remote configuration, why not document that origin is the default instead of requiring it to be explicit?

This is an option, although as someone who is going to have to answer support questions regarding "why is the default for this experiment `origin` instead of the default I set on my git branch", my preference would be to continue requiring the explicit argument

Reply:
@dberenbaum 
> Alternatively, I think we could reconsider having a default remote.

I think having a default remote for the `dvc exp push` is reasonable, both `dvc push` and `git push` have default remotes.
How about adding it to the `dvc.config`, it is more manageable. 

And if a default remote for `dvc exp push` is setted, the only env arguments here to control the auto push process is `DVC_EXP_CHECKPOINT_PUSH:bool`.

I don't think users will directly use auto push without using `dvc exp push` previously. If they are familiar with `dvc exp push`, and knows how to set up a default remote. The auto push configuration is quite a simple one.